### PR TITLE
Optimize payload builder using a shared compressor buffer

### PR DESCRIPTION
**Pull request by gh123man**
### What does this PR do?

In scenarios where lots of metrics are sent that are compressed as streamed json, the agent spends a lot of time allocating buffers to compress the payloads. This PR introduces shared compressor bufferes for JSON streamed payloads. A result of sharing bufferes means we need to lock access to the compressor since it can be called concurrently. In my testing this does not appear to cause a bottleneck and the benefits of sharing buffers seems to outweigh the loss of a fully concurrent compressor workflow. 

In the future we should re-design the serializer to support shared bufferes in more places while preserving concurrency/maximum throughput. 

Benchmarking this was tricky because there are 2 variables that affect performance: 
1. The size of the payload
2. The throughput/frequency of payloads of a given size

So I broke the benchmarks down into small, medium, and large payloads with a number of passes to perform (to simulate sending different volumes of different sized payloads). See the results below comparing using shared buffers vs the original buffer allocation strategy:

Large payloads (100000 series, increasing number of passes)
```
benchmark                                old ns/op        new ns/op        delta
BenchmarkJSONStreamSharedLarge1-8        154595181        146649406        -5.14%
BenchmarkJSONStreamSharedLarge10-8       1482717623       1439636010       -2.91%
BenchmarkJSONStreamSharedLarge100-8      15256261064      14623200928      -4.15%
BenchmarkJSONStreamSharedLarge1000-8     153517217124     149535568716     -2.59%

benchmark                                old allocs     new allocs     delta
BenchmarkJSONStreamSharedLarge1-8        120            116            -3.33%
BenchmarkJSONStreamSharedLarge10-8       1196           1156           -3.34%
BenchmarkJSONStreamSharedLarge100-8      11818          11416          -3.40%
BenchmarkJSONStreamSharedLarge1000-8     118025         114029         -3.39%

benchmark                                old bytes      new bytes      delta
BenchmarkJSONStreamSharedLarge1-8        9325260        5363436        -42.48%
BenchmarkJSONStreamSharedLarge10-8       91252496       49660720       -45.58%
BenchmarkJSONStreamSharedLarge100-8      871406424      413160784      -52.59%
BenchmarkJSONStreamSharedLarge1000-8     8655743392     4039970496     -53.33%
```

Medium sized payloads (10000 series, increasing number of passes)
```
benchmark                            old ns/op        new ns/op        delta
BenchmarkJSONStreamSharedMed1-8         15525793         15068957         -2.94%
BenchmarkJSONStreamSharedMed10-8        155500742        149138299        -4.09%
BenchmarkJSONStreamSharedMed100-8       1550220929       1501759959       -3.13%
BenchmarkJSONStreamSharedMed1000-8      15555692979      15056454807      -3.21%
BenchmarkJSONStreamSharedMed10000-8     157920881222     152996603687     -3.12%

benchmark                           old allocs     new allocs     delta
BenchmarkJSONStreamSharedMed1-8        38             34             -10.53%
BenchmarkJSONStreamSharedMed10-8       382            342            -10.47%
BenchmarkJSONStreamSharedMed100-8      3826           3415           -10.74%
BenchmarkJSONStreamSharedMed1000-8     38086          34022          -10.67%

benchmark                            old bytes       new bytes      delta
BenchmarkJSONStreamSharedMed1-8         3244234         966864         -70.20%
BenchmarkJSONStreamSharedMed10-8        32442269        9668644        -70.20%
BenchmarkJSONStreamSharedMed100-8       323416088       94692872       -70.72%
BenchmarkJSONStreamSharedMed1000-8      3213014240      905065824      -71.83%
BenchmarkJSONStreamSharedMed10000-8     32108972424     9008794992     -71.94%
```

Small Payloads (100 series, increasing number of passes)
```
benchmark                                 old ns/op      new ns/op      delta
BenchmarkJSONStreamSharedSmall1-8         324270         312559         -3.61%
BenchmarkJSONStreamSharedSmall10-8        3178791        3150656        -0.89%
BenchmarkJSONStreamSharedSmall100-8       35486475       30562818       -13.87%
BenchmarkJSONStreamSharedSmall1000-8      324149840      315537200      -2.66%
BenchmarkJSONStreamSharedSmall10000-8     3521839000     3129099801     -11.15%

benchmark                                old allocs     new allocs     delta
BenchmarkJSONStreamSharedSmall1-8        38             34             -10.53%
BenchmarkJSONStreamSharedSmall10-8       380            340            -10.53%
BenchmarkJSONStreamSharedSmall100-8      3805           3405           -10.51%
BenchmarkJSONStreamSharedSmall1000-8     38058          34053          -10.52%

benchmark                                 old bytes      new bytes      delta
BenchmarkJSONStreamSharedSmall1-8         842147         819526         -2.69%
BenchmarkJSONStreamSharedSmall10-8        8421460        8195268        -2.69%
BenchmarkJSONStreamSharedSmall100-8       84214650       81952646       -2.69%
BenchmarkJSONStreamSharedSmall1000-8      842146284      819526338      -2.69%
BenchmarkJSONStreamSharedSmall10000-8     8421448424     8195223808     -2.69%
```

Single pass - increasing payload size:
```
benchmark                         old ns/op       new ns/op       delta
BenchmarkJSONStream1-8            200343          177310          -11.50%
BenchmarkJSONStream10-8           217911          209190          -4.00%
BenchmarkJSONStream100-8          369933          330251          -10.73%
BenchmarkJSONStream1000-8         2155447         1856775         -13.86%
BenchmarkJSONStream10000-8        19013872        15774750        -17.04%
BenchmarkJSONStream100000-8       173723580       158715148       -8.64%
BenchmarkJSONStream1000000-8      1578004938      1720906460      +9.06%
BenchmarkJSONStream10000000-8     15058265781     16867166875     +12.01%

benchmark                        old allocs     new allocs     delta
BenchmarkJSONStream1-8           38             34             -10.53%
BenchmarkJSONStream10-8          38             34             -10.53%
BenchmarkJSONStream100-8         38             34             -10.53%
BenchmarkJSONStream1000-8        38             34             -10.53%
BenchmarkJSONStream10000-8       38             34             -10.53%
BenchmarkJSONStream100000-8      120            116            -3.33%
BenchmarkJSONStream1000000-8     970            966            -0.41%

benchmark                         old bytes     new bytes     delta
BenchmarkJSONStream1-8            826933        818645        -1.00%
BenchmarkJSONStream10-8           827028        818725        -1.00%
BenchmarkJSONStream100-8          842147        819527        -2.69%
BenchmarkJSONStream1000-8         1115311       827555        -25.80%
BenchmarkJSONStream10000-8        3256753       969840        -70.22%
BenchmarkJSONStream100000-8       9444090       5363440       -43.21%
BenchmarkJSONStream1000000-8      46578056      46479672      -0.21%
BenchmarkJSONStream10000000-8     382625160     383100200     +0.12%
```

### Describe your test plan

Tested in real environments. 
Benchmarks

QA: Most of the QA for this card will be looking at telemetry to ensure: 
1. Metrics are still coming through
2. There are no performance bottlenecks 
3. There is some measurable performance improvement 

Some (expected) things to look out for: 
- Slight drop in CPU usage overall
- less erratic RSS
- a slight (but not major) increase in static heap size due to the longer living buffers

I expect to co-QA this card during the release so whoever is assigned this please feel free to reach out to me.

### CODE REVIEW

File: pkg/serializer/jsonstream/payload_builder.go [hunk 1]
```diff
@@ -27,13 +36,26 @@ var jsonConfig = jsoniter.Config{
 // use multiple PayloadBuilders for different sources.
```
Comment (olivielpeau) [inline pkg/serializer/jsonstream/payload_builder.go:36]: docstring should be updated with `shareAndLockBuffers`'s behavior

**Review by olivielpeau — COMMENTED** [review]
Thanks! I think it's also worth adding unit tests that ensure that the new logic produces the same compressed output as the old logic.

The benchmarks are interesting: the gains are not as significant as I initially expected, probably because the compressor still needs to allocate new buffers in which it returns the compressed payloads.

Let's measure during QA if overall this change increases or decreases the heap.

(please update the PR description with a more detailed QA plan, and the expected effects of this PR on the "static" heap of the Agent)

File: pkg/serializer/jsonstream/payload_builder.go [hunk 2]
```diff
@@ -9,14 +9,23 @@ package jsonstream
...
 
 	"github.com/DataDog/datadog-agent/pkg/forwarder"
 	"github.com/DataDog/datadog-agent/pkg/serializer/marshaler"
+	"github.com/DataDog/datadog-agent/pkg/telemetry"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
+var tlmCompressorLocks = telemetry.NewGauge(
+	"payload_builder",
+	"blocking_goroutines",
```
[the remark this answers is not in the document]
Comment (gh123man) [reply pkg/serializer/jsonstream/payload_builder.go]: Added a total latency measurement. LMK if it looks good to you. I did some spot testing in `/debug/vars` to verify the counters are working.

**Review by olivielpeau — APPROVED** [review]
LGTM!

Let's closely track how this behaves in QA 👀

**Pull request merged**